### PR TITLE
Create a Java API for setting colspan

### PR DESCRIPTION
**Pull request by svipal**
Create a Java API for setting colspan (https://github.com/vaadin/vaadin-form-layout-flow/issues/68)

This a PR for review purposes.

To start off with, I have never used vaadin before and am really surprised at how smooth this is. Kudos.

How does this look so far ? Is this what you had in mind ? 

Questions : 

- Is it fine to silently put to 1 the colspans that are set to less than 1 using the API or would an exception throw be preferable ?

- Is the way I make sure the colspan attribute is set or valid acceptable ? 

- I understand the integration testing process but I'm not really sure as to what I should test ?... 
A thing I can see that could be tested from the client side is the size of the columns, but apparently there's already a test regarding columns and it is disabled because it fails when resizing.
It probably wouldn't work with definite colspans either.

- why is there indentation in the demo ? There isn't any in the code.

image

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-form-layout-flow/84)

### CODE REVIEW

File: vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java [hunk 1]
```diff
@@ -211,6 +211,62 @@ public FormLayout() {
...
+     *
+     * @param component
+     *            the component to modify
+     *
+     * @param colspan
+     *            the desired colspan for the component
+     *
+     */
+    public void setColspan(Component component, int colspan) {
+    	String str_colspan = "";
```
[the remark this answers is not in the document]
Comment (svipal) [reply vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java]: fixed in latest commit

File: vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java [hunk 2]
```diff
@@ -211,6 +211,62 @@ public FormLayout() {
...
+     * Gets the colspan of a component, if any. if none is set, returns 1
+     *
+     * @param component
+     *            the component whose colspan is retrieved
+     *
+     *
+     */
+    
+    public int getColspan(Component component) {
+    	String str_colspan = component.getElement().getAttribute("colspan");
```
[the remark this answers is not in the document]
Comment (svipal) [reply vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java]: fixed in latest commit

File: vaadin-form-layout-flow-demo/src/main/java/com/vaadin/flow/component/formlayout/demo/FormLayoutView.java [hunk 3]
```diff
@@ -57,6 +57,7 @@ public void initView() {
         createFormLayoutWithItems();
         createFormLayoutWithBinder();
         createCompositeLayout();
+        createLayoutHandleColspans();
```
Comment (pleku) [inline vaadin-form-layout-flow-demo/src/main/java/com/vaadin/flow/component/formlayout/demo/FormLayoutView.java]: Actually, I think this demo could be shown already as the third demo after `createFormLayoutWithItems()`. Please move it there.

File: vaadin-form-layout-flow-demo/src/main/java/com/vaadin/flow/component/formlayout/demo/FormLayoutView.java [hunk 4]
```diff
@@ -329,4 +330,45 @@ private void createCompositeLayout() {
 
         addCard("Using form layout inside a composite", layout);
     }
+    
+    private void createLayoutHandleColspans() {
+        // begin-source-example
+        // source-example-heading: Handling columns and colspans in a layout
+    	FormLayout colLayout = new FormLayout();
```
Comment (pleku) [inline vaadin-form-layout-flow-demo/src/main/java/com/vaadin/flow/component/formlayout/demo/FormLayoutView.java]: >    	FormLayout colLayout = new FormLayout();
I would name it just `layout` or use full name `columnLayout`.

File: vaadin-form-layout-flow-demo/src/main/java/com/vaadin/flow/component/formlayout/demo/FormLayoutView.java [hunk 5]
```diff
@@ -329,4 +330,45 @@ private void createCompositeLayout() {
...
+    	//Setting the desired responsive steps for the columns in the layout
```
Comment (pleku) [inline vaadin-form-layout-flow-demo/src/main/java/com/vaadin/flow/component/formlayout/demo/FormLayoutView.java]: Just very small format thing, to add whitespace after the `//`
```suggestion
    	// Setting the desired responsive steps for the columns in the layout
```

File: vaadin-form-layout-flow-demo/src/main/java/com/vaadin/flow/component/formlayout/demo/FormLayoutView.java [hunk 6]
```diff
@@ -329,4 +330,45 @@ private void createCompositeLayout() {
...
+        lastName.setPlaceholder("Last Name");
+        TextField email = new TextField();
+        email.setPlaceholder("Email");
+        TextField nickname = new TextField();
+        nickname.setPlaceholder("Username");
+        TextField website = new TextField();
+        website.setPlaceholder("Link to personal website");
+        TextField description = new TextField();
+        description.setPlaceholder("Enter a short description about yourself");
+        //Adding the components to it.
```
Comment (pleku) [inline vaadin-form-layout-flow-demo/src/main/java/com/vaadin/flow/component/formlayout/demo/FormLayoutView.java]: The comment is IMO unnecessary, and just an empty line before it is enough 
```suggestion

```

File: vaadin-form-layout-flow-demo/src/main/java/com/vaadin/flow/component/formlayout/demo/FormLayoutView.java [hunk 7]
```diff
@@ -329,4 +330,45 @@ private void createCompositeLayout() {
...
+        colLayout.add(firstName, lastName,  nickname, email, website); 
+        //you can set the desired column span for them individually.
```
Comment (pleku) [inline vaadin-form-layout-flow-demo/src/main/java/com/vaadin/flow/component/formlayout/demo/FormLayoutView.java]:
```suggestion
        // you can set the desired column span for them individually.
```

File: vaadin-form-layout-flow-demo/src/main/java/com/vaadin/flow/component/formlayout/demo/FormLayoutView.java [hunk 8]
```diff
@@ -329,4 +330,45 @@ private void createCompositeLayout() {
...
+        colLayout.setColspan(website, 2);
+        //or just set it while adding them :
```
Comment (pleku) [inline vaadin-form-layout-flow-demo/src/main/java/com/vaadin/flow/component/formlayout/demo/FormLayoutView.java]:
```suggestion
        // you can set the desired column span for them individually.
```

Comment (svipal) [reply vaadin-form-layout-flow-demo/src/main/java/com/vaadin/flow/component/formlayout/demo/FormLayoutView.java]: Woops missed the typo there. fixed in latest commit.

File: vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java [hunk 9]
```diff
@@ -211,6 +211,65 @@ public FormLayout() {
     public FormLayout(Component... components) {
         add(components);
     }
+    
+    /**
+     * Sets the colspan of the given component's element.
+     * You can directly add components with the wanted colspan with
+     * {@link #add(Component)}.
```
Comment (pleku) [inline vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java]: >	     * {@link #add(Component)}.
This should probably link to `#add(Component, int)` instead ?

File: vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java [hunk 10]
```diff
@@ -211,6 +211,65 @@ public FormLayout() {
...
+     *
+     * @param component
+     *            the component to modify
```
Comment (pleku) [inline vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java]:
```suggestion
	     *            the component to set the colspan for, not {@code null}
```

File: vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java [hunk 11]
```diff
@@ -211,6 +211,65 @@ public FormLayout() {
...
+     *
+     * @param colspan
+     *            the desired colspan for the component
+     *
+     */
+    public void setColspan(Component component, int colspan) {
+    	String strColspan = "";
```
Comment (pleku) [inline vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java]:
```suggestion
                Objects.requireNonNull(component, "component cannot be null");
	    	String strColspan = "";
```

File: vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java [hunk 12]
```diff
@@ -211,6 +211,65 @@ public FormLayout() {
...
+    	if (colspan < 1) {
+    		strColspan = "1";
```
Comment (pleku) [inline vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java]: >    	if (colspan < 1) {
>    		strColspan = "1";

This is a side effect and not mentioned in the javadocs. If you do this, you need to mention that what happens for values <1. Another option is to throw `IllegalArgumentException`

File: vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java [hunk 13]
```diff
@@ -211,6 +211,65 @@ public FormLayout() {
...
+    	} else {
+    		strColspan=String.valueOf(colspan);
+    	}
+    	component.getElement().setAttribute("colspan", strColspan );
+    }
+    
+    /**
+     * Adds a component with the desired colspan.
+     *
```
Comment (pleku) [inline vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java]: Could add here a mention that "This method is a shorthand for calling #add(Component) and #setColspan(Component, int)"

File: vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java [hunk 14]
```diff
@@ -211,6 +211,65 @@ public FormLayout() {
...
+     */
+    public void add(Component component, int colspan) {
+    	add(component);
+    	setColspan(component, colspan);
+    	
+    }
+    
+    /**
+     * Gets the colspan of a component, if any. if none is set, returns 1
```
Comment (pleku) [inline vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java]:
```suggestion
     * Gets the colspan of the given component. If none is set, returns 1.
```

File: vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java (hunk 9 above)
Comment (svipal) [reply vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java]: You're right.

File: vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java [hunk 15]
```diff
@@ -211,6 +211,65 @@ public FormLayout() {
...
+     *
+     * @param component
+     *            the component whose colspan is retrieved
+     *
+     *
+     */
+
```
Comment (pleku) [inline vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java]: extra whitespace
```suggestion

```

File: vaadin-form-layout-flow/src/test/java/com/vaadin/flow/component/formlayout/tests/FormLayoutTest.java [hunk 16]
```diff
@@ -27,4 +27,35 @@ public void getResponsiveSteps_noInitialSteps_emptyListIsReturned() {
         FormLayout layout = new FormLayout();
         Assert.assertTrue(layout.getResponsiveSteps().isEmpty());
     }
+    
+    @Test
```
Comment (pleku) [inline vaadin-form-layout-flow/src/test/java/com/vaadin/flow/component/formlayout/tests/FormLayoutTest.java]: Missing a test that actually validates that the colspan is passed to the component's element correctly, e.g. checks `componentWithColspan.getElement().getAttribute("colspan")`

**Comment by pleku** [comment]
Also you should rebase your changes on top of master branch, since there are conflicts.

**Comment by svipal** [comment]
You're welcome; No problem. 
Will do !

File: vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java (hunk 12 above)
Comment (svipal) [reply vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java]: Mentioned it in latest commit.

File: vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java (hunk 11 above)
Comment (svipal) [reply vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java]: Done in latest commit.

File: vaadin-form-layout-flow/src/test/java/com/vaadin/flow/component/formlayout/tests/FormLayoutTest.java (hunk 16 above)
Comment (svipal) [reply vaadin-form-layout-flow/src/test/java/com/vaadin/flow/component/formlayout/tests/FormLayoutTest.java]: Resolved in latest commit

**Comment by svipal** [comment]
Should be all right !

**Comment by ZheSun88** [comment]
the validation failed on 
```
Compilation failure
/opt/agent/work/21749182162872d3/vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java:[231,9] cannot find symbol
  symbol:   variable Objects
  location: class com.vaadin.flow.component.formlayout.FormLayout
```

**Comment by svipal** [comment]
That's my bad - when I added the constraint  Objects.requireNonNull() I forgot to import java.util.Objects. Fixed.

File: vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java [hunk 17]
```diff
@@ -30,6 +31,7 @@
 import org.jsoup.helper.StringUtil;
 
 import com.vaadin.flow.component.html.Label;
+import org.jsoup.helper.StringUtil;
```
[the remark this answers is not in the document]
Comment (svipal) [reply vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java]: Arf

[the remark this answers is not in the document]
Comment (svipal) [reply vaadin-form-layout-flow/src/main/java/com/vaadin/flow/component/formlayout/FormLayout.java]: Leftover from master merge

**Pull request merged**